File: src/features/typing/data/go/path_sum_iii.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, targetSum int) int {
	count := map[int]int{0: 1}
	var dfs func(node *TreeNode, presum int) int
	dfs = func(node *TreeNode, presum int) int {
		if node == nil {
			return 0
		}
		presum += node.Val
		total := count[presum-targetSum]
		count[presum]++
		total += dfs(node.Left, presum) + dfs(node.Right, presum)
		count[presum]--
		return total
	}
	return dfs(root, 0)
}

File: src/features/typing/data/java/path_sum_iii.java
import java.util.HashMap;
import java.util.Map;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int pathSum(TreeNode root, int targetSum) {
        Map<Integer, Integer> count = new HashMap<>();
        count.put(0, 1);
        return dfs(root, 0, targetSum, count);
    }

    private int dfs(TreeNode node, int presum, int targetSum, Map<Integer, Integer> count) {
        if (node == null) {
            return 0;
        }
        presum += node.val;
        int total = count.getOrDefault(presum - targetSum, 0);
        count.put(presum, count.getOrDefault(presum, 0) + 1);
        total += dfs(node.left, presum, targetSum, count) + dfs(node.right, presum, targetSum, count);
        count.put(presum, count.get(presum) - 1);
        return total;
    }
}
